src/listener.js

```diff
@@ -34,20 +33,26 @@ class Listener {
 
   listen() {
     if (this.archiveBucketSettings) {
-      this.replaying = true;
+      this.processingEvent = true;
       eventReplayer.replayEvents(this.archiveBucketSettings, this._bucketEventHandler, this.logger).then(() => {
         // This happens after *all* events from the bucket have been replayed
-        this.replaying = false;
+        this.processingEvent = false;
       }).catch(this.logger.error); // Replay never finishes if it fails
     }
     return (req, res) => {
+      if (this.processingEvent) {
+        this.logger.info('Rejecting an event because another one is already in progress.');
+        res.status(503).json({ error: 'Currently processing another event. Try again soon.' });
+        return Promise.resolve();
+      }
+      this.processingEvent = true;
```

Comment:
Probably a silly question, but what's to stop us setting `this.processingEvent` to the event ID, and then checking if it matches above?

Reply:
Interesting. Are you thinking we could set it to the event ID and check against that, and then we would allow parallel event processing as long as it's not the *same* event?

I think that could possibly work, but then instead of a single event ID we'd need to store an array or set of IDs, and check if any of them match. So not too hard, but a little more complex.

I haven't really thought about the implications of that though. My intention was always to make it so that events are processed one at a time, in order. I feel more comfortable with that because then we should have fewer edge races with race conditions etc. What do you think?

Reply:
I guess that makes sense - an event stream is linear, so you want the events processed linearly.

I wonder how well it will scale (both in terms of events taking longer times to process, and in terms of higher number of events coming through).

Reply:
The scaling question is an interesting one. I'm not sure how it would affect performance. JavaScript is inherently single-threaded, so we can't have true parallelism regardless. I guess where it would slow things down is if we're processing one event which takes a while because it does something slow and async (like sending email), then we can't process any other events in the mean time.

Not sure what the impact will be in practice. The only way to find out would be to throw a crapload of events at a real environment and see how it performs I guess! Maybe we should try that!